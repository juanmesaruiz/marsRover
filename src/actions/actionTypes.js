export const ROVER_CLEAR = 'ROVER@CLEAR';
export const ROVER_CLEAR_POSITION_LOG = 'ROVER@CLEAR_POSITION_LOG';
export const ROVER_DELETE_OBSTACLE = 'ROVER@DELETE_OBSTACLE';
export const ROVER_NEW_INSTRUCTIONS = 'ROVER@NEW_INSTRUCTIONS';
export const ROVER_NEW_MOVE = 'ROVER@NEW_MOVE';
export const ROVER_SET_GRID = 'ROVER@SET_GRID';
export const ROVER_SET_OBSTACLES = 'ROVER@SET_OBSTACLES';
export const ROVER_SET_POSITION = 'ROVER@SET_POSITION';
