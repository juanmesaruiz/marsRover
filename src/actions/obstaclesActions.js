import { ROVER_SET_OBSTACLES } from './actionTypes';

export const createObstacles = obstacles => ({
  type: ROVER_SET_OBSTACLES,
  payload: {
    obstacles,
  },
});
