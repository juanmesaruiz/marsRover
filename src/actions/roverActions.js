import { ROVER_CLEAR, ROVER_NEW_MOVE, ROVER_SET_POSITION } from './actionTypes';

export const roverClear = () => ({
  type: ROVER_CLEAR,
});

export const roverSetPosition = position => ({
  type: ROVER_SET_POSITION,
  payload: {
    position,
  },
});

export const roverNewKeyboardMove = ({ code, grid, obstaclesCoordinates }) => ({
  type: ROVER_NEW_MOVE,
  payload: {
    code,
    grid,
    obstaclesCoordinates,
  },
});
