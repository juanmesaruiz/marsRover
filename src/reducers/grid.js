import { ROVER_CLEAR } from '../actions/actionTypes';

const defaultState = { x: 50, y: 25 };

const grid = (state = defaultState, action) => {
  switch (action.type) {
    case ROVER_CLEAR:
      return defaultState;

    default:
      return state;
  }
};

export default grid;
