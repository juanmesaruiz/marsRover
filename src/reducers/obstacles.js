import { ROVER_CLEAR, ROVER_SET_OBSTACLES } from '../actions/actionTypes';

const defaultState = [];

const grid = (state = defaultState, action) => {
  switch (action.type) {
    case ROVER_CLEAR:
      return defaultState;

    case ROVER_SET_OBSTACLES: {
      const { obstacles } = action.payload;
      return [...state, ...obstacles];
    }

    default:
      return state;
  }
};

export default grid;
