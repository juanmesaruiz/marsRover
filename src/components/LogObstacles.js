import React, { useCallback } from 'react';
import { connect } from 'react-redux';
import styled from 'styled-components';

import { deleteObstacle } from '../actions/obstaclesActions';
import { getObstacles } from '../reducers';

import Button from './library/Button';

const LogObstacles = ({ className, deleteObstacle, obstacles }) => {
  const handleDeleteObstacle = useCallback(
    obstacle => deleteObstacle(obstacle),
    [deleteObstacle]
  );

  return (
    <div className={className}>
      <h5 className="title">Total obstacles {obstacles.length}:</h5>
      <div className="positions-container">
        {obstacles.map(({ x, y }) => (
          <p key={`${x}-${y}`}>
            X {x}, Y {y}
            <Button
              className="delete-button"
              onClick={() => handleDeleteObstacle({ x, y })}
            >
              Delete
            </Button>
          </p>
        ))}
      </div>
    </div>
  );
};

const StyledLogObstacles = styled(LogObstacles)`
  .delete-button {
    float: right;
    margin-right: 20px;
  }

  .title {
    margin: 25px 0 0;
  }

  .positions-container {
    height: 225px;
    overflow: auto;
  }
`;

const mapStateToProps = state => ({
  obstacles: getObstacles(state),
});

export default connect(mapStateToProps, { deleteObstacle })(StyledLogObstacles);
