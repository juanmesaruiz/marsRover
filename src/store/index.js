import { applyMiddleware, compose, createStore } from 'redux';
import thunk from 'redux-thunk';

import reducers from '../reducers';

const configureStore = () => {
  const initialState = {};
  const enhancers = [];
  const devToolsExtension = window.__REDUX_DEVTOOLS_EXTENSION__;

  if (typeof devToolsExtension === 'function') {
    enhancers.push(devToolsExtension({ name: '- ROVER MARS -' }));
  }

  const composedEnhancers = compose(applyMiddleware(thunk), ...enhancers);

  const store = createStore(reducers, initialState, composedEnhancers);

  return { store };
};

export default configureStore;
