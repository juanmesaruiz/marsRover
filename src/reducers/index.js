import { combineReducers } from 'redux';

import grid from './grid';
import obstacles from './obstacles';
import rover from './rover';

const reducers = {
  grid,
  obstacles,
  rover,
};

export const getGrid = state => state.grid;
export const getObstacles = state => state.obstacles;
export const getRover = state => state.rover;

export default combineReducers(reducers);
