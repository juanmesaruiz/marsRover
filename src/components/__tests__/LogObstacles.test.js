import React from 'react';
import ReactDOM from 'react-dom';
import { Provider } from 'react-redux';
import { fireEvent, render } from '@testing-library/react';
import 'jest-styled-components';

import { deleteObstacle } from '../../actions/obstaclesActions';
import { getObstacles } from '../../reducers';

import LogObstacles from '../LogObstacles';

import store from '../../store/__mocks__/mockStore';
import state from '../../store/__mocks__/mockState';

jest.mock('../../actions/obstaclesActions');
jest.mock('../../reducers');

describe('LogObstacles test suite', () => {
  const component = (
    <Provider store={store}>
      <LogObstacles />
    </Provider>
  );

  const setup = () => render(component);

  beforeEach(() => {
    deleteObstacle.mockReturnValue({
      type: 'deleteObstacle',
    });
    getObstacles.mockReturnValue(state.obstacles);
  });

  afterEach(jest.clearAllMocks);

  it('Renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(component, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('Should match snapshot', () => {
    const { container } = setup();
    expect(container).toMatchSnapshot();
  });

  it('Should fire deleteObstacle action', () => {
    const { getAllByText } = setup();
    fireEvent.click(getAllByText('Delete')[0]);

    expect(deleteObstacle).toHaveBeenCalledTimes(1);
  });
});
