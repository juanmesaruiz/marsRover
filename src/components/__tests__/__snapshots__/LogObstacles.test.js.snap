// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LogObstacles test suite Should match snapshot 1`] = `
.c1 {
  cursor: pointer;
}

.c0 .delete-button {
  float: right;
  margin-right: 20px;
}

.c0 .title {
  margin: 25px 0 0;
}

.c0 .positions-container {
  height: 225px;
  overflow: auto;
}

<div>
  <div
    class="c0"
  >
    <h5
      class="title"
    >
      Total obstacles 
      5
      :
    </h5>
    <div
      class="positions-container"
    >
      <p>
        X 
        0
        , Y 
        3
        <button
          class="c1 delete-button"
        >
          Delete
        </button>
      </p>
      <p>
        X 
        2
        , Y 
        1
        <button
          class="c1 delete-button"
        >
          Delete
        </button>
      </p>
      <p>
        X 
        4
        , Y 
        4
        <button
          class="c1 delete-button"
        >
          Delete
        </button>
      </p>
      <p>
        X 
        3
        , Y 
        3
        <button
          class="c1 delete-button"
        >
          Delete
        </button>
      </p>
      <p>
        X 
        1
        , Y 
        4
        <button
          class="c1 delete-button"
        >
          Delete
        </button>
      </p>
    </div>
  </div>
</div>
`;
